src/main/kotlin/org/eln2/mc/Extensions.kt

```diff
@@ -0,0 +1,949 @@
+}
+
+@ServerOnly
+fun ServerLevel.destroyPart(part: Part<*>, dropPart: Boolean) {
+    val pos = part.placement.pos
+
+    val multipart = this.getBlockEntity(pos)
+        as? MultipartBlockEntity
+
+    if (multipart == null) {
+        LOG.error("Multipart null at $pos")
+
+        return
+    }
+
+    val saveTag = CompoundTag()
+
+    multipart.breakPart(part, saveTag)
+
+    if(dropPart) {
+        val itemEntity = ItemEntity(
+            this,
+            pos.x.toDouble(),
+            pos.y.toDouble(),
+            pos.z.toDouble(),
+            Part.createPartDropStack(part.id, saveTag)
+        )
+
+        this.addFreshEntity(itemEntity)
+    }
+
+    if (multipart.isEmpty) {
+        this.destroyBlock(pos, false)
+    }
+}
+
+inline fun <reified TEntity : BlockEntity> Level.constructMenu(
+    pos: BlockPos,
+    player: Player,
+    crossinline title: (() -> Component),
+    factory: ContainerFactory<TEntity>,
+): InteractionResult {
+
+    if (!this.isClientSide) {
+        val entity = this.getBlockEntity(pos) as? TEntity
+            ?: return InteractionResult.FAIL
+
+        val containerProvider = object : MenuProvider {
+            override fun getDisplayName(): Component {
+                return title()
+            }
+
+            override fun createMenu(
+                pContainerId: Int,
+                pInventory: Inventory,
+                pPlayer: Player,
+            ): AbstractContainerMenu {
+                return factory.create(
+                    pContainerId,
+                    pInventory,
+                    pPlayer,
+                    entity
+                )
+            }
+        }
+
+        NetworkHooks.openScreen(player as ServerPlayer, containerProvider, entity.blockPos)
+        return InteractionResult.SUCCESS
+    }
+
+    return InteractionResult.SUCCESS
+}
+
+inline fun <reified TEntity : BlockEntity> Level.constructMenu(
+    pos: BlockPos,
+    player: Player,
+    crossinline title: (() -> Component),
+    crossinline factory: ((Int, Inventory, ItemStackHandler) -> AbstractContainerMenu),
+    crossinline accessor: ((TEntity) -> ItemStackHandler),
+): InteractionResult {
+
+    return this.constructMenu<TEntity>(pos, player, title) { id, inventory, _, entity ->
+        factory(id, inventory, accessor(entity))
+    }
+}
+
+fun Level.getDataAccess(pos: BlockPos): DataNode? {
+    return ((this.getBlockEntity(pos) ?: return null)
+        as? DataEntity ?: return null)
+        .dataNode
+}
+
+inline fun <reified T : Cell> Level.getCellOrNull(mb: MultiblockManager, cellPosId: BlockPos): T? {
+    val entity = this.getBlockEntity(mb.txIdWorld(cellPosId)) as? CellBlockEntity
+        ?: return null
+
+    return entity.cell as? T
+}
+
+inline fun <reified T : Cell> Level.getCell(mb: MultiblockManager, cellPosId: BlockPos): T =
+    getCellOrNull(mb, cellPosId) ?: error("Cell was not present")
+
+private const val LIBAGE_SET_EPS = 0.001
```

Comment:
There's a bunch of libage extensions here (below) that might make sense as part of libage? Might want to move them, in time.

Reply:
that, or move to libage.kt?